core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/AbstractFederationConnection.java

```diff
@@ -434,10 +437,18 @@ private TupleExpr optimize(TupleExpr parsed, Dataset dataset, BindingSet binding
 		new QueryMultiJoinOptimizer().optimize(query, dataset, bindings);
 		// new FilterOptimizer().optimize(query, dataset, bindings);
 
-		new EmptyPatternOptimizer(members).optimize(query, dataset, bindings);
+		// prepare bloom filters
+		RepositoryBloomFilter defaultBloomFilter = new AccurateRepositoryBloomFilter(includeInferred);
+		Map<Repository, RepositoryBloomFilter> bloomFilters = federation.getBloomFilters();
+		java.util.function.Function<Repository, RepositoryBloomFilter> bloomFilterFunction = c -> bloomFilters.getOrDefault(
```

Comment:
If you change this function to `<Function, Optional<RepositoryBloomFliter>>` (implementation would be `c -> Optional.ofNullable(bloomFilters.get...))` then in the places where you now use `MoreObjects` you can simply do `firstFilter.or(secondFilter)`.